### ISSUE TITLE
Screen reader keeps losing focus

### ISSUE THREAD

**Issue by rkingett** [ChoicescriptIDE/main#14]
I am a blind author who wanted to give this a try, but I had trouble with keeping my screen reader to stay focused. When I would tab around the main area, even, I would get stuck in one place or the other. I would open the keyboard commands link only to get stuck in that help frame. I tried using the dropbox version but it had a lot of ubnlabeled buttons so I had no idea what I was pressing.

I would be more than happy to answer questions or show you, via audio, what is happening.

**Comment by CareyJWilliams** [comment]
Hi @rkingett, screen reader support is definitely something that needs work. If you're happy to help with that, it would be very much appreciated. To start with your current issue though, what operating system and screen reader software are you using?

**Comment by rkingett** [comment]
I have no programming experience at all, but I use NVDA, on a windows 10 
machine. I am running the latest build, and can test, and report back 
when needed, and the like.

My public email is kingettr@gmail.com if you would like to converse that 
way.

The browser version is a little bit better focus wise, but it has a LOT 
of unlabeled buttons the desktop version does not have.

**Comment by CareyJWilliams** [comment]
They are actually the same app, same code base etc., the problem is that different browsers can vary quite a bit in how they handle accessibility attributes. What browser is that you are using primarily?

I have started adding in some of the missing labels, hopefully we can iron those out fairly quickly. I think the trickier part will be making sure that the entire application is easily navigable (on all platforms). Beyond tab and shift-tab, do you make much use of other controls? I know mac's voiceover can bring up a list of headings etc.

**Comment by rkingett** [comment]
yeah, I use a lot of heading navigation, button navigation, link navigation, and otherwise. I am using FireFox when I try and run the browser version... I just can't quite tell how the program is layed out visually. I am guessing it is in frames or something? In certain windows documentation, hitting F6 jumps to various frames, like left pannel, right, etc, and then I can navigate in there as if I am looking at a web page... 

There is an unlabeled edit box somewhere on every pannel, even the help, and I can't quite figure out what that is for. Is that where I enter the text? 

Does it look like an IM conversation, where there is a text box and then the code displays at the top? I can't really use the spell check, either, because NVDA does not echo back anything at all when I try to type in whatever edit box that is... 

I think it was, is, the text area box because when I force my cursor out of the box, because it won't tab out without a few mouse clicks randomly, I hear a pannel that sounds like... 

if I typed, for example... 

you walk down the street... 

I hear... 

1. youwalkdownthestreet. 

text sounds like it has no space, so what is that? Is it possible you can send me the keyboard commands as a text document so I can look at them without getting stuck in that help window? pannel? 

I am sorry I can't give much feedback, currently but it is extremely dificult to tell where I am, and that is just by tabbing around, exploring. 

Let me know if you have any more questions!

**Comment by rkingett** [comment]
Which version would be better, testing wise? The dropbox version or the windows desktop version? How can I know if there is a new build for me to test out?

**Comment by CareyJWilliams** [comment]
Sorry for the delayed response. I'm fairly confident the desktop versions get more traffic, though both *should* (in theory) be very similar, at least across Chrome/Chromium.

The main application is split into three panes, there is a project and scene list, an editor window (which displays the scene selected in the list) and a tab panel, which contains five or so different selectable tabs providing additional functionality, such as a user dictionary, the game window, issue tracking etc.

The desktop version has an updater, but I think the prompts could be made more noticeable to screenreaders. The web version will just update when I update it. I'll post when there's an update in the thread on choiceofgames.com. Because the web version doesn't have update "channels" I may or may not update it immediately, it will depend on the scope of the changes.

---

This is all very helpful, I think a lot falls to me to just make some basic workable changes, and then, once you feel it's even remotely useable, we can work to tweak and tailor the details a bit more.

**Comment by rkingett** [comment]
I think that working on the desktop version would be best for now since it has an updater. Is any of the documentation online somewhere like a wiki or something? If Not could you send the documenattion in an email or something?

I would also be more than happy to write and edit the accessibility wiki on this page, as well.

I am curious, how is this different than Notepad++? Does it have autocomplete options, like, if I type...

*pgla

it will male 2 lines of...

*label

*page_break

I can't wait to try this out! Thank you so very much for being open to this. It will make a lot of other blind coders happy...

One issue I am having with Notepad++ is I keep forgetting what label is in what scene, and, of course, the tab structure of all the choices. Can this program assist with the *if command too? 

Why do so many people like using this better than notepad++? What do you use?

I am a newbie, and keep forgetting the *set commands I have in a scene too... could this have highlighting, with line numbers?

**Comment by CareyJWilliams** [comment]
Hi, yes you can just can just access the web version's help docs directly here: https://choicescriptide.github.io/web/help/

A wiki would be great, once we have some ground work down.

If you mean does it have macros, custom codes that can be expanded into other things, no (not yet).
Sadly there's no extra processing or monitoring of project state (which labels in scenes etc) either. 

What do you mean by assist with the *if command?

I programmed CSIDE in atom and Notepad++, for ChoiceScript I use the ChoiceScript IDE. The difference is astronomical. No one is intending for CSIDE to replace Notepad++, but the fact is, CSIDE was designed from the ground up specifically for ChoiceScript, Notepad++ was not. The IDE allows you to run the the game side-by-side, track game data, better manage projects, automatically update your local copy of ChoiceScript and much, much more. If you're just looking for an editor, CSIDE won't ever compete with Notepad++, but if you're looking to improve your ChoiceScript game development process, then that's where CSIDE really shines.

All the choicescript is highlighted in CSIDE, but I'm not sure that's what you mean?

**Comment by rkingett** [comment]
I am sure this will become clearer to me once I have a usable product, but I am just curious if there is a way to make the screen reader cursor jump directly to the visually highlighted bit?

I found the resources file in the help page, but how do I open it?

Something that would help with alignment and otherwise... is, for example, in microsoft wworrd, when I hit tab, NVDA says, 1.5 inches from left margin...

But maybe the smart formatting already does that? I am trying to get a grasp of how I can test when the time comes. I have subscribed to this repository too so I can get updates...

I was thinking in terms of the *elsif command. could the program assist with lining up the *elsif and *else commands to the *if commands?

I have been looking for an accessibility checker for notepad++ for you but I could not find a plugin. Sorry about that!

I am not sure if this would help at all, but below is a link to an accessible gamebook authoring tool. It is not as feature rich, nor does it help in any way with Choice Script but it can give you an idea of the structure of a fully accessible program.

I currently use [this program](http://www.kaldobsky.com/audiogames/Darkgrue.zip) to outline my game.

**Comment by CareyJWilliams** [comment]
Visually highlighted? I know MAC's voiceover can follow the cursor, but this is all down to the screen reader, not the IDE. The help docs are just a series of web pages, there's nothing to directly download. You can also find the raw markdown files here in this repository under source/help/topics.

Again I think space/tab counting is likely screen reader specific? I certainly have no idea how I'd alert the SR of that information from CSIDE. Auto indent should help, but sadly auto dedent can only do so much, as it's impossible to know when a user has finished under an *if or *choice (i.e. there's no "end if" command) , it's up to them to manually dedent.

I will take a look thank you.

**Comment by CareyJWilliams** [comment]
@rkingett  Just to let you know I am making some progress on this. If you'd be interested in helping via testing, do you think you would be able to build the application from the Github source? Thinking that might be easier than a version release, whilst we iron out the details.

**Comment by rkingett** [comment]
I only know how to write in markdown and choice script, but yeahh, would like to test and see what you are working on. Is it one of the masters? If so how would I get to it?

**Comment by CareyJWilliams** [comment]
You wouldn't need to code anything, just download (or clone) the Github repository, and enter some build commands. You would require NodeJS, NPM and a global Grunt installed though. Maybe I can just send you a build directly. Would you be alright replacing a file in a given directory?

**Comment by rkingett** [comment]
Oh certainly! I can do that, replace a file in a given direcctory, yes. My email is kingettr@gmail.com. If it is more than one file, could you send it as a zip file?

**Comment by CareyJWilliams** [comment]
It'll be a .nw file, which is essentially a renamed .zip. You'll need to navigate to CSIDE's install directory (usually in your %APPDATA% directory) and replace the "package.nw" file there (after backing it up!).

I don't have a build ready yet, but when I do, I'll pass it along. Thanks!

**Comment by rkingett** [comment]
I would like to check in on a branch to test? Is there one that works better with screen readers yet?

**Comment by CareyJWilliams** [comment]
Hi @rkingett, yes, you can download and build from the "accessibility" branch to gain all the current accessibility updates. I imagine this is still far from finished though. Any feedback you can give would be incredibly helpful.

**Comment by CareyJWilliams** [comment]
If you're struggling to clone and build, give me a shout here and I will get you a prebuilt copy somehow.

**Comment by rkingett** [comment]
I found [version 2](https://github.com/ChoicescriptIDE/main/projects/1) but can't quite figure out how to, well, build. I am running windows 10. Could you send me a zip file? My email is kingett r at gmale dot com.

**Comment by CareyJWilliams** [comment]
@rkingett Sure, no problem. I'll get one built this week. What platform are you on?

**Comment by rkingett** [comment]
I run windows 10. 64 bit.

**Comment by CareyJWilliams** [comment]
Hey @rkingett, sorry for the delay on this. I'm slowly chipping away at this. What I ideally would like to sort out is screen reader compatibility for updates. That way, once you have a working copy, I can push you any improvements without a need for reinstallation.

**Comment by CareyJWilliams** [comment]
https://github.com/ChoicescriptIDE/main/issues/40
